httpserver/bin/basic_writer_server.dart

```diff
@@ -9,36 +9,37 @@
 import 'dart:io';
 import 'dart:convert';
 
-void main() {
+main() async {
+  var server =
+      await HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 4049);
+  await for (var req in server) {
+    ContentType contentType = req.headers.contentType;
+    BytesBuilder builder = new BytesBuilder();
 
-  HttpServer.bind(InternetAddress.LOOPBACK_IP_V4, 4049).then((server) {
-    server.listen((req) {
+    if (req.method == 'POST' &&
+        contentType != null &&
+        contentType.mimeType == 'application/json') {
+      try {
+        await for (var buffer in req) builder.add(buffer);
 
-      ContentType contentType = req.headers.contentType;
-      BytesBuilder builder = new BytesBuilder();
-
-      if (req.method == 'POST' &&
-          contentType != null &&
-          contentType.mimeType == 'application/json') {
-        req.listen((buffer) {
-          builder.add(buffer);
-        }, onDone: () {
-          // Write to a file, get the file name from the URI.
-          String jsonString = UTF8.decode(builder.takeBytes());
-          String filename = req.uri.pathSegments.last;
-          new File(filename).writeAsString(jsonString,
-              mode: FileMode.WRITE).then((_) {
-            Map jsonData = JSON.decode(jsonString);
-            req.response.statusCode = HttpStatus.OK;
-            req.response.write('Wrote data for ${jsonData['name']}.');
-            req.response.close();
-          });
-        });
-      } else {
-        req.response.statusCode = HttpStatus.METHOD_NOT_ALLOWED;
-        req.response.write("Unsupported request: ${req.method}.");
+        // Write to a file, get the file name from the URI.
+        String jsonString = UTF8.decode(builder.takeBytes());
```

Comment:
In other places we transform the stream using UTF8.decoder before looping over it.
I guess this one is different because it was just written by someone else.

I guess we could just do:
    var jsonString = await req.transform(UTF8.decoder).join();
instead of having a loop at all?